pkg/services/statefetcher/statefetcher.go

```diff
@@ -0,0 +1,364 @@
+	GetConfig() config.Blockchain
+	HeaderHeight() uint32
+	AddContractStorageData(kv []statesync.KeyValue, syncHeight uint32, expectedRoot util.Uint256) error
+	GetLastStoredKey() []byte
+}
+
+type kvPair struct {
+	statesync.KeyValue
+	err error
+}
+
+// Service fetches contract storage state from NeoFS.
+type Service struct {
+	basic             neofs.BasicService
+	isActive          atomic.Bool
+	isShutdown        atomic.Bool
+	cfg               config.NeoFSStateFetcher
+	stateSyncInterval uint32
+	lastStateIndex    uint32
+	lastStateOid      oid.ID
+
+	chain Ledger
+	log   *zap.Logger
+	wg    sync.WaitGroup
+
+	shutdownCallback func()
+}
+
+// New creates a new Service instance.
+func New(chain Ledger, cfg config.NeoFSStateFetcher, stateSyncInterval int, logger *zap.Logger, shutdownCallback func()) (*Service, error) {
+	if !cfg.Enabled {
+		return &Service{}, nil
+	}
+	if cfg.Timeout <= 0 {
+		cfg.Timeout = neofs.DefaultTimeout
+	}
+	if cfg.StateAttribute == "" {
+		cfg.StateAttribute = neofs.DefaultStateAttribute
+	}
+	if cfg.KeyValueBatchSize <= 0 {
+		cfg.KeyValueBatchSize = neofs.DefaultKvBatchSize
+	}
+
+	basic, err := neofs.New(cfg.NeoFSService)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create service: %w", err)
+	}
+	sfs := Service{
+		basic:             basic,
+		log:               logger,
+		cfg:               cfg,
+		chain:             chain,
+		shutdownCallback:  shutdownCallback,
+		stateSyncInterval: uint32(stateSyncInterval),
+	}
+	if sfs.stateSyncInterval <= 0 {
+		sfs.stateSyncInterval = config.DefaultStateSyncInterval
+	}
+	var (
+		containerID  cid.ID
+		containerObj container.Container
+	)
+	sfs.basic.Ctx, sfs.basic.CtxCancel = context.WithCancel(context.Background())
+	if err = sfs.basic.Pool.Dial(context.Background()); err != nil {
+		sfs.isShutdown.CompareAndSwap(false, true)
+		return nil, fmt.Errorf("failed to dial NeoFS pool: %w", err)
+	}
+
+	err = containerID.DecodeString(sfs.cfg.ContainerID)
+	if err != nil {
+		sfs.isShutdown.CompareAndSwap(false, true)
+		return nil, fmt.Errorf("failed to decode container ID: %w", err)
+	}
+
+	err = sfs.basic.Retry(func() error {
+		containerObj, err = sfs.basic.Pool.ContainerGet(sfs.basic.Ctx, containerID, client.PrmContainerGet{})
+		return err
+	})
+	if err != nil {
+		sfs.isShutdown.CompareAndSwap(false, true)
+		return nil, fmt.Errorf("failed to get container: %w", err)
+	}
+	containerMagic := containerObj.Attribute("Magic")
+	if containerMagic != strconv.Itoa(int(sfs.chain.GetConfig().Magic)) {
+		sfs.isShutdown.CompareAndSwap(false, true)
+		return nil, fmt.Errorf("container magic mismatch: expected %d, got %s", sfs.chain.GetConfig().Magic, containerMagic)
+	}
+	sfs.isShutdown.CompareAndSwap(true, false)
+	return &sfs, nil
+}
+
+// LastStateIndex returns the last state index from NeoFS.
+func (sfs *Service) LastStateIndex() (uint32, error) {
+	if sfs.lastStateIndex != 0 {
+		return sfs.lastStateIndex, nil
+	}
+	startIndex := uint32(1)
+	for {
```

Comment:
Infinite loops are not the best solution. Can we use context bound to `sfs` to track the node interruption initiated by user at the moment of object search?